![GitHub Release](https://img.shields.io/github/v/release/KirklandLab/CutandTag_ReplicatePeak_Analysis)
![GitHub Release Date](https://img.shields.io/github/release-date/KirklandLab/CutandTag_ReplicatePeak_Analysis)
![GitHub repo size](https://img.shields.io/github/repo-size/KirklandLab/CutandTag_ReplicatePeak_Analysis)
![GitHub last commit](https://img.shields.io/github/last-commit/KirklandLab/CutandTag_ReplicatePeak_Analysis)
![GitHub contributors](https://img.shields.io/github/contributors/KirklandLab/CutandTag_ReplicatePeak_Analysis)
![GitHub Downloads (all assets, all releases)](https://img.shields.io/github/downloads/KirklandLab/CutandTag_ReplicatePeak_Analysis/total)
![GitHub commits since latest release](https://img.shields.io/github/commits-since/KirklandLab/CutandTag_ReplicatePeak_Analysis/latest)
[![DOI](https://zenodo.org/badge/898608902.svg)](https://doi.org/10.5281/zenodo.15232319)
[![License: Apache 2.0](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)

# CutandTag_ReplicatePeak_Analysis

![ReplicatePeaks](/images/replicatePeaks.png)
*(Image generated with DALL-E. OpenAI, 2024: Scientific data visualization: Replicate peak analysis in bioinformatics)*

--- 

## 1) Project Description

**CutAndTag_ReplicatePeak_Analysis** is a Snakemake pipeline designed to perform downstream analysis on processed Cut-and-Tag sequencing data. Rather than starting from raw FASTQ reads, this pipeline starts with aligned BAM files, focusing on the identification of reproducible peaks, the generation of consensus peak sets, and the visualization of overlaps and signal distributions across multiple samples or experimental conditions.

+ Note: If you are starting from raw FASTQ files, use the upstream [CutandTag_Alignment_QC](https://github.com/KirklandLab/CutandTag_Alignment_QC) pipeline first. That pipeline handles the initial data processing steps providing quality control analysis, alignment, and multiple signal tracks.

### **Key Features**

+ **Peak Calling with MACS2 on Individual and Merged Samples**:
  + Each replicate is processed independently using MACS2 to call peaks. In parallel, biological replicates defined by the `set` column in `samples.csv` are merged at the BAM level, and a second MACS2 peak calling is performed on the merged BAM to capture reproducible signal across replicates.

+ **Reproducible Consensus Peak Sets**:
  + A consensus peak set is generated by comparing merged peaks to individual replicate peaks. Peaks are retained only if supported by a minimum number of replicates specified in the `config.yml` (e.g., 2 out of 3), ensuring biological reproducibility while reducing noise.

+ **Signal Preservation Through Merged Read Support**:
  + Reads are extracted directly from the merged BAM preserving read-level signal. This enables more accurate representation of genomic signal in consensus peak regions.

+ **Consensus Peak Conversion to Visualization Formats**:
  + Consensus peak regions are used to extract reads and generate sorted BAM and BigWig files.
  + These outputs are optimized for genome browser visualization and downstream comparative analysis.

+ **Euler Plots of Replicate and Condition Overlaps**:
  + Euler diagrams are created to visualize how consensus peaks arise from replicate overlaps within a set.
  + Additional Euler plots show condition-level peak set relationships for each mark, providing clear insight into reproducibility and condition-specific differences. Plot color and size can be set in the `config.yml`.
  + **Default colors**: `#4477AA` (blue), `#228833` (green), `#CCBB44` (mustard/golden yellow)
  + **Extended pallete** (for >3 samples): `#EE6677` (reddish coral), `#AA3377` (violet/purple), `#66CCEE` (light cyan/sky blue)

+ **Midpoint and Overlap Quantification**:
  + Consensus peaks are processed to determine their midpoints and identify overlaps across all sets. This provides a standardized reference point for comparing distributions and regional enrichments.

+ **High-Resolution Heatmaps of Signal Intensity**:
  + The pipeline generates heatmaps centered on consensus peak midpoints. These plots visualize the distribution and intensity of signal across replicates, conditions, or sample sets, making trends in coverage and peak architecture easy to interpret.
  + Heatmaps are customizable in the `config.yml` along with the ability to toggle them on and off to minimize the runtime for large datasets.
  + The `sampleOrder` section in the `config.yml` sets the sample order for heatmaps, with the first sample determining the order of the BED file in the heatmap.

+ **Optional Customizable Heatmap Generation**:
  + A helper script is included in scripts/make_heatplot.sh for visualizing BigWig signals over defined BED regions.
  + Instructions for customizing and running the script are provided as comments.

---

## 2) Intended Use Case  

This pipeline is ideal for researchers who have already processed their Cut-and-Tag data through preliminary steps such as quality control, read alignment, and signal track generation. These preprocessing steps can be completed using the upstream [CutandTag_Alignment_QC](https://github.com/KirklandLab/CutandTag_Alignment_QC) workflow.  

After obtaining high-quality aligned BAM files, this pipeline can be used to:  

+ Identify **reproducible peaks** across biological replicates
+ Generate **consensus peak sets** for each condition or mark
+ Visualize **overlaps across replicates and experimental conditions** using Euler diagrams
+ Produce **heatmaps** showing signal intensity around consensus peak midpoints

By integrating this two-step approach, you ensure a robust and reproducible end-to-end workflow for your Cut-and-Tag sequencing experiments.

---

## 3) Dependencies and Configuration

Tool versions and parameters (e.g., genome size, MACS2 q-values, minimum number of overlapping samples for consensus peaks, paths to executables) are defined in the `config/config.yml` file. By default, the `config.yml` is set up for mouse (mm10). Running human (hg38) samples requires changing these values to match the hg38 parameters, which are already provided in `config.yml` as comments. Specific tool versions help maintain reproducibility and ensure that the pipeline runs consistently across different computing environments.  

**Key Configuration Fields**

+ `genome`: genome ID (`mm`, `hs`)
+ `effective_genome_size`: required by MACS2 and DeepTools
+ `chrom_sizes`: chromosome sizes file path
+ `min_overlap`: number of replicates needed to retain a peak
+ `heatmap_on`: toggle for heatmap generation
+ `sampleOrder`: controls order of samples in heatmap

**Changing Genome**

+ For hg38:
  + `genome: "hs"`
  + `effective_genome_size: 2913022398`
  + `chrom_sizes: resources/hg38.chrom.sizes`

+ For mm10:
  + `genome: "mm"`
  + `effective_genome_size: 2730871774`
  + `chrom_sizes: resources/mm10.chrom.sizes`

---

##  4) Tools & Modules  

The pipeline relies on bioinformatics tools, including:  

- **MACS2**: for peak calling
- **Bedtools / Samtools**: for peak and alignment format conversions
- **Deeptools** for coverage / matrix computation as well as for generating heatmaps
- **R + Bioconductor**: for merging peaks, checking overlap, creating midpoints, and creating **Euler** diagrams
- **Python**: for performing workflow logic 

---

## 5) Example Data
A compact, pre-processed dataset (provided by Jacob Kirkland) aligned to mm10 is included in this repository to quickly test the pipeline and validate that your environment is set up correctly. This small example replicates the pipeline’s key steps from peak calling through to final visualization.  

---

## 6) Explanation of `samples.csv`  

**IMPORTANT**: Always check the `config/samples.csv` with your sample name, BAM file locations, set, mark, and condition before running.

**Required Columns**  
+ `sample`: unique name for sample  
  + used for individual sample output filenames and to identify samples in consensus peak euler plot  
+ `bam`: path to BAM file  
  + Sorted BAM files are recommended, but other BAM formats are also supported  
+ `set`: sample group for consensus peaks  
  + **{mark}_{condition}**  
  + used to name consensus peak files downstream. All samples with the same **Set** name will be combined to generate a consensus peak  
  + **MUST MATCH MARK AND CONDITION** (e.g. H3K27ac_Control, H3K27ac_C, H3K27ac_T, etc.)  
+ `mark`: indicates the histone mark name or other feature captured  
  + all samples with the same mark name will be grouped for subsequent control vs treatment plots  
  + **MUST MATCH MARK NAME IN SET** (e.g. H3K27ac, H3K27me3, H3K4me3, etc.)  
+ `condition`: specifies the experimental condition  
  + these values are used to distinguish groups in control vs treatment overlap plots for each mark  
  + **MUST MATCH CONDITION NAME IN SET** (e.g. C, T, Control, treatment, etc.)  

**Example**
```csv
sample,bam,set,mark,condition
H3K27ac_rep1_C,resources/H3K27ac_rep1_C_PB042_chr1_1_10Mb.bam,H3K27ac_C,H3K27ac,C
H3K27ac_rep1_T,resources/H3K27ac_rep1_T_PB045_chr1_1_10Mb.bam,H3K27ac_T,H3K27ac,T
H3K27ac_rep2_C,resources/H3K27ac_rep2_C_PB052_chr1_1_10Mb.bam,H3K27ac_C,H3K27ac,C
H3K27ac_rep2_T,resources/H3K27ac_rep2_T_PB055_chr1_1_10Mb.bam,H3K27ac_T,H3K27ac,T
H3K27ac_rep3_C,resources/H3K27ac_rep3_C_PB062_chr1_1_10Mb.bam,H3K27ac_C,H3K27ac,C
H3K27ac_rep3_T,resources/H3K27ac_rep3_T_PB065_chr1_1_10Mb.bam,H3K27ac_T,H3K27ac,T
```

---

## 7) Output Overview

| Category                   | Output Location                                         |
|---------------------------|---------------------------------------------------------|
| MACS2 Peaks               | `results/peaks/` (`.narrowPeak` files)                  |
| Consensus Peaks           | `results/peaks/consensus/`                              |
| Sorted BAM / BigWig       | `results/alignment/`                                    |
| Euler Diagrams            | `results/plots/euler/`                                  |
| Heatmaps                  | `results/plots/heatmaps/`                               |
| Midpoint & Overlap Files  | `results/peaks/midpoints/`                              |

---

## 8) Example Output Plots  

Below are some example plots generated by the pipeline.  

| 1. **H3K4me3_C_eulerPlot**                                                            | 2. **H3K4me3_T_eulerPlot**                                                                 |
| :-----------------------------------------------------------------------------------: | :----------------------------------------------------------------------------------------: |
| <img src="/images/H3K4me3_C_eulerPlot.png" width="300">                               | <img src="/images/H3K4me3_T_eulerPlot.png" width="300">                                    |
| *Euler plot of individual sample overlaps in Condition C. Illustrates consensus peaks*| *Euler plot of individual sample overlaps in Condition T. Illustrates consensus peaks*     |

| 3. **H3K4me3_eulerPlot**                                                                                                                                                           |
| :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
| <img src="/images/H3K4me3_eulerPlot.png" width="300">                                                                                                                              |
| *Euler plot showing consensus peak overlap of conditions in each mark. H3K4me3 shown.*                                                                                             |

| 4. **HeatPlots_All_Samples**                                                          | 5. **HeatPlots_Unique_Samples**                                                            |
| :-----------------------------------------------------------------------------------: | :----------------------------------------------------------------------------------------: |
| <img src="/images/HeatPlots_All_Samples.png" width="300">                             | <img src="/images/HeatPlots_Unique_Samples.png" width="300">                               |
| *Heatmap showing signal distributions for all samples midpoints.*                     | *Heatmap showing signal distributions for all samples midpoints separated by sample*       |


# 9) Instructions to run on Slurm managed HPC
9A. Download version controlled repository
```
wget https://github.com/KirklandLab/CutandTag_ReplicatePeak_Analysis/releases/download/v1.2.1/CutandTag_ReplicatePeak_Analysis-1.2.1.tar.gz
tar -xzf CutandTag_ReplicatePeak_Analysis-1.2.1.tar.gz
rm CutandTag_ReplicatePeak_Analysis-1.2.1.tar.gz
cd CutandTag_ReplicatePeak_Analysis-1.2.1
```
9B. Load modules
```
module purge
module load slurm python/3.10 pandas/2.2.3 numpy/1.22.3 matplotlib/3.7.1
```
9C. Modify samples and config file
```
vim config/samples.csv
vim config/config.yml
```
9D. Dry Run
```
snakemake -npr
```
9E. Run on HPC with config.yml options
```
sbatch --wrap="snakemake -j 999 --use-envmodules --latency-wait 60 --cluster-config config/cluster_config.yml --cluster 'sbatch -A {cluster.account} -p {cluster.partition} --cpus-per-task {threads} --time {cluster.time} --mem {cluster.mem} --output {cluster.output} --job-name {cluster.name}'"
```

---

## 10) Citation

+ **Boyd, K.A.** (2025). *CutandTag_ReplicatePeak_Analysis: A reproducible Snakemake workflow for consensus peak calling and visualization in Cut-and-Tag data*. *Zenodo*. https://doi.org/10.5281/zenodo.15232319  
+ [![DOI](https://zenodo.org/badge/898608902.svg)](https://doi.org/10.5281/zenodo.15232319)

---

## 11) Authorship & Contributions

+ **Kevin A. Boyd** – Designed and implemented the Snakemake workflow for a Slurm-managed HPC environment, modularized the pipeline structure, implemented all processing steps, integrated peak consensus method, designed plots, and created the documentation.  
+ **Jacob Kirkland** – Principal Investigator; provided project direction, conceptual guidance, experimental data, and validation of biological logic.  

This workflow was developed under the guidance of Jacob Kirkland as part of a COBRE-funded collaborative effort. While the pipeline was built specifically for use within the Kirkland Lab, it is broadly applicable to Cut-and-Tag data analysis in other research settings.  

---

## 12) License

This project is licensed under the **Apache 2.0**. See the [LICENSE](LICENSE) file for details.  

[![License: Apache 2.0](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)

